k8s/upgrade-job/src/helm/upgrade.rs

```diff
@@ -0,0 +1,192 @@
+        client::{HelmReleaseClient, HelmReleaseClientBuilder},
+        values::generate_values_args,
+    },
+};
+use regex::Regex;
+use serde_yaml::Value;
+use snafu::ResultExt;
+use std::{fs, path::PathBuf};
+
+/// This is the helm chart variant of the helm chart installed in the cluster.
+/// The PRODUCT may be installed using either of these options, but never both.
+#[derive(Clone, Default)]
+pub(crate) enum HelmChart {
+    #[default]
+    Umbrella,
+    Core,
+}
+
+/// This is a builder for the Helm chart upgrade.
+#[derive(Default)]
+pub(crate) struct HelmUpgradeBuilder {
+    chart_variant: HelmChart,
+    release_name: String,
+    namespace: String,
+    umbrella_chart_dir: Option<PathBuf>,
+    core_chart_dir: Option<PathBuf>,
+}
+
+impl HelmUpgradeBuilder {
+    /// This creates a default instance of the HelmUpgradeBuilder.
+    pub(crate) fn default() -> Self {
+        Self {
+            umbrella_chart_dir: None,
+            core_chart_dir: None,
+            ..Default::default()
+        }
+    }
+
+    /// This is the same as the default() method, but the name of the method may be more mppropriate
+    /// for builder uses.
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    /// This is a builder option to add the Namespace of the helm chart to be upgrade.
+    #[must_use]
+    pub(crate) fn with_namespace<J>(mut self, ns: J) -> Self
+    where
+        J: ToString,
+    {
+        self.namespace = ns.to_string();
+        self
+    }
+
+    /// This is a builder option to add the release name of the helm chart to be upgraded.
+    #[must_use]
+    pub(crate) fn with_release_name<J>(mut self, release_name: J) -> Self
+    where
+        J: ToString,
+    {
+        self.release_name = release_name.to_string();
+        self
+    }
+
+    /// This is a builder option to set the directory path of the Umbrella helm chart CLI option.
+    #[must_use]
+    pub(crate) fn with_umbrella_chart_dir(mut self, dir: Option<PathBuf>) -> Self {
+        self.umbrella_chart_dir = dir;
+        self
+    }
+
+    /// This is a builder option to set the directory path of the Umbrella helm chart CLI option.
+    #[must_use]
+    pub(crate) fn with_core_chart_dir(mut self, dir: Option<PathBuf>) -> Self {
+        self.core_chart_dir = dir;
+        self
+    }
+
+    /// This adds buiilds the HelmUpgrade object.
+    pub(crate) fn build(mut self) -> Result<HelmUpgrade> {
+        // Generate HelmReleaseClient.
+        let client = HelmReleaseClientBuilder::new()
+            .with_namespace(self.namespace.clone())
+            .build();
+        // Get HelmReleaseElement object for the release specified in CLI options.
+        let chart = client.release_info(self.release_name.clone())?.chart();
+
+        // Define regular expression to pick out the chart name from the
+        // <chart-name>-<chart-version> string.
+        let umbrella_chart_regex = format!(r"^({UMBRELLA_CHART_NAME}-[0-9]+\.[0-9]+\.[0-9]+)$");
```

Comment:
Seems this won't support prereleases?

Reply:
Added support for pre-releases on this regex, also on the version lock in `crate::upgrade::path::is_validI()`.